batch-deletion/main.py

```diff
@@ -0,0 +1,167 @@
+
+# List Containing pending iteration objects
+PendingList = queue.Queue()
+UsageSummary = queue.Queue()
+
+StopProcessing = False
+InitialIterationDone = False
+TotalContainer = 0
+
+MaxThreadCount = 8
+IterationRunning = 0
+IteratorLock = threading.Lock()
+
+IteratorPoolList = []
+
+
+def thread_handler(thrId):
+    # print("Child " + str(thrId) + " started")
+    global InitialIterationDone
+    global IterationRunning
+    global MaxThreadCount
+
+    Done = False
+
+    total_path_processed = 0
+    while not Done:
+        try:
+            queue_item = PendingList.get(timeout=3)
+        except:
+            if PendingList.empty() and IterationRunning == 0:
+                Done = True
+            continue
+
+        IteratorLock.acquire()
+        IterationRunning += 1
+        IteratorLock.release()
+
+        # print("Iterating path : " + queue_item.Container + ":" + queue_item.ListPath)
+
+        total_path_processed += 1
+
+        dir_count = 0
+        file_count = 0
+        total_size = 0
+        if queue_item.Type == "Block":
+            blob_list = azure.list_storage_blobs(queue_item.Container, queue_item.ListPath)
+            blob_path_list = []
+            for blob in blob_list:
+                blob_info = azure.get_blob_info(blob)
+                if blob_info["dir"]:
+                    dir_count += 1
+                    PendingList.put(QueueItem(Container=queue_item.Container, Type="Block", ListPath=blob_info["name"]))
+                else:
+                    file_count += 1
+                    total_size += blob_info["size"]
+                    blob_path_list.append(blob_info["name"])
+                blob_info.clear()
+
+            if len(blob_path_list) > 0:
+                UsageSummary.put(DeletionQueue(Container=queue_item.Container, BlobList=blob_path_list))
+
+        PendingList.task_done()
+
+        IteratorLock.acquire()
+        IterationRunning -= 1
+        IteratorLock.release()
+
+    # print("Child " + str(thrId) + " exited : Dir " + str(total_path_processed))
+    IteratorLock.acquire()
+    MaxThreadCount -= 1
+    IteratorLock.release()
+
+
+
+def delete_batch():
+    Done = False
+    while not Done:
+        try:
+            usage_item = UsageSummary.get(block=False, timeout=3)
+        except:
+            if MaxThreadCount == 0:
+                Done = True
+            continue
+
+        print(usage_item.BlobList)
+        azure.delete_blob_batch(usage_item.Container, usage_item.BlobList)
+        UsageSummary.task_done()
+
+
+# ----------------------- Main processing --------------------------
+def main(cli_options):
+    global MaxThreadCount
+
+    if cli_options['connection_string']:
+        azure.set_connection_string(cli_options['connection_string'])
+    else:
+        conn_str = os.getenv("AZURE_STORAGE_CONNECTION_STRING")
+        azure.set_connection_string(conn_str)
+
+    if cli_options['parallel_factor']:
+        MaxThreadCount = int(cli_options['parallel_factor'])
```

Comment:
nit: The cli option `parallel_factor` is a little misleading. If I did not want to run in parallel I might just naturally pass 0 (trying to only run 1 thread)

Reply:
If `parallel_factor` is not passed, we use default `MaxThreadCount=8`. The whole purpose of this Batch Deletion code we wrote was to implement it using Multithreading. If multithreading is not required, user can pass `parallel_factor=1`